### PR TITLE
ETX contract deployment

### PR DESCRIPTION
**Pull request by jdowning100**
ETXs can now deploy smart contracts by adding a smart contract to the ETXData field (in InternalToExternal tx) and setting the To field to the destination chain's contextual "zero" address. The "zero" address is computed by placing the destination location byte prefix at the front of the address and padding the other 19 bytes with zeroes. This PR builds upon changes made in https://github.com/dominant-strategies/go-quai-libp2p/pull/60 and fixes https://github.com/dominant-strategies/go-quai/issues/1313

Edit: Tested by deploying a contract with internal to external tx from zone-0-0 to zone-0-1, contract deployed successfully and I was able to query the state of the contract.

### CODE REVIEW

**Review by wizeguyy — APPROVED** [review]

**Comment by wizeguyy** [comment]
awesome sauce. thanks for this!

File: common/address.go [hunk 1]
```diff
@@ -212,7 +212,7 @@ func (a Address) Value() (driver.Value, error) {
 // Location looks up the chain location which contains this address
 func (a Address) Location(nodeLocation Location) *Location {
 	if a.inner == nil {
-		return &nodeLocation
+		return &Location{} // nil address does not have a location
```
[the remark this answers is not in the document]
Comment (jdowning100) [reply common/address.go:215]: We can return nil or panic, I think either are fair

[the remark this answers is not in the document]
Comment (wizeguyy) [reply common/address.go:215]: lets panic. I left a similar comment on your first PR

[the remark this answers is not in the document]
Comment (jdowning100) [reply common/address.go]: Fixed and rebased

**Review by wizeguyy — APPROVED** [review]

**Review by wizeguyy — APPROVED** [review]

**Review by wizeguyy — APPROVED** [review]

**Pull request merged**